### ISSUE TITLE
Question: How to extend Streaming Sortformer beyond 4 speakers?

### ISSUE THREAD

**Issue by tsugumi-sys** [NVIDIA/NeMo#14546]
Hi, thank you so much for releasing Streaming Sortformer!

I tested it on an audio clip with 6 speakers, and I noticed that only 4 speaker labels (spk0–spk3) are produced. I understand this is an expected limitation of the current architecture, not a bug.

That said, I’d like to ask for some advice:

### What I tried

- Used Streaming Sortformer to get time-aligned speaker segments.
- Extracted MFCC features from each segment.
- Applied online k-means clustering to assign speaker IDs.

This gave me not-bad accuracy for distinguishing 6 speakers. I didn’t pre-train the clustering model, so I assume accuracy could improve with more sophisticated embeddings or training.

### My questions

- Do you think combining Streaming Sortformer with lightweight clustering/identification methods (like the one above) is a reasonable direction?
- Or is there a better approach to handle more than 4 speakers in real time?

Sorry if this is not the best place to ask, but I’d really appreciate any suggestions or guidance!

Thanks again for your great work.

**Comment by ipmedenn** [comment]
Hi @tsugumi-sys,

Thank you for your interest in Streaming Sortformer and for sharing your detailed feedback!
We're glad to hear you're putting it to the test.

Your approach of combining our model with a lightweight clustering method is a clever idea for extending its capabilities.
You are correct that the 4-speaker limit is an architectural constraint.
Here are a few thoughts and suggestions on your questions.

You've correctly identified a path to handle more speakers.
However, when running inference on audio with more than 4 speakers, two main types of errors can occur:
- **Speaker Confusion**: The speech of a fifth (or more) speaker is mistakenly assigned to one of the four tracked speakers (e.g., spk0-spk3). Your clustering method can effectively help correct these misattributions.
- **Missed Speech**: The speech from additional speakers is not assigned to any of the four tracked speakers and is dropped. Unfortunately, your post-processing clustering can't recover this lost data.

To improve performance for your use case, we suggest two possible directions.

1. Use better frame-level representations for clustering
To improve your current setup, we recommend using the outputs from the pre-encoder module of Sortformer instead of raw MFCC features. These pre-encoder embeddings are much richer and more speaker-discriminative, which should significantly boost the accuracy of your k-means clustering.

2. Adopt the EEND-VC Paradigm
For a more robust solution, you could try the End-to-End Neural Diarization with vector clustering (EEND-VC) approach described in this paper https://arxiv.org/abs/2105.09040. This approach is well-suited for handling an unknown and large number of speakers.
This involves:
    - Splitting the audio into shorter chunks (10-20s)
    - Using a local diarization model like Streaming Sortformer on each chunk to identify speech segments for up to 4 speakers at a time.
    - Then, for each resulting segment, extracting a high-quality speaker embedding using a dedicated model (we suggest looking into TitaNet https://huggingface.co/nvidia/speakerverification_en_titanet_large).
    - Finally, performing a global clustering on these embeddings to handle the total number of speakers in the audio.

I hope this guidance is helpful! Please don't hesitate to reach out with any more questions. Thanks again for your valuable feedback.

**Comment by tsugumi-sys** [comment]
Hi @ipmedenn 

Thank you so much for your very kind and detailed response — it really helps me a lot!
Since I need to perform diarization in a real-time setting, I’ll focus on trying option 1 first and experiment with those richer representations instead of raw MFCCs! I'll also check option2 after that.

Really appreciate your support and guidance — it’s been invaluable for me.
Please close this issue. Thanks!

**Comment by SGD2718** [comment]
Tip: Use the fast conformer embeddings, NOT the pre-encoder embeddings. The pre-encoder embeddings capture local speech features and do not cluster nicely.

<img width="2684" height="1480" alt="Image" src="https://github.com/user-attachments/assets/ab70202a-fa9f-4c72-9f52-0161484285cc" />
Fast Conformer Embeddings

<img width="2683" height="1480" alt="Image" src="https://github.com/user-attachments/assets/720d880b-e465-4649-b867-87f86b5b050c" />
Pre-Encoder Embeddings

**Comment by karynaur** [comment]
@SGD2718 Interesting observation. Can you share the code that was used to generate these plots?

**Comment by SGD2718** [comment]
Update: after further experimentation, it appears that the embeddings seem to encode the speaker arrival order indices more than they encode speaker identities. 

Those were UMAP plots of the embeddings from a CoreML port of the model (since I'm an Apple user). You'd need to make a custom Torch wrapper to do that without the CoreML stuff. 

The code is in single_file_coreml_plots.py:
https://github.com/Audivize-AI/Streaming-Sortformer-Conversion

**Comment by Hanminh** [comment]
> Tip: Use the fast conformer embeddings, NOT the pre-encoder embeddings. The pre-encoder embeddings capture local speech features and do not cluster nicely.
> 
> <img alt="Image" width="2000" height="1480" src="https://private-user-images.githubusercontent.com/48599511/531377529-ab70202a-fa9f-4c72-9f52-0161484285cc.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU1ODA1NjgsIm5iZiI6MTc3NTU4MDI2OCwicGF0aCI6Ii80ODU5OTUxMS81MzEzNzc1MjktYWI3MDIwMmEtZmE5Zi00YzcyLTlmNTItMDE2MTQ4NDI4NWNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDA3VDE2NDQyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjZTcxNTAyODk3NzUwMTJhN2RmYjBjYzE4NDliNmQxMmI0N2M1MmQ2NzExN2I0ZWQ1MzM1Y2Y5N2M0ZWI4MjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Oor04aoKdNpeIHFxi4Yc0IfTo7VjU9AXQr_51nqycII"> Fast Conformer Embeddings <img alt="Image" width="2000" height="1480" src="https://private-user-images.githubusercontent.com/48599511/531377545-720d880b-e465-4649-b867-87f86b5b050c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU1ODA1NjgsIm5iZiI6MTc3NTU4MDI2OCwicGF0aCI6Ii80ODU5OTUxMS81MzEzNzc1NDUtNzIwZDg4MGItZTQ2NS00NjQ5LWI4NjctODdmODZiNWIwNTBjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDA3VDE2NDQyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlOTkwOTRkNWQ5Y2I1YTg0MWViOTdmOGI0NjVjMTQ1Y2EyYjI4MWI5NDk3NjMzNzYyZjVlYjA4MzgxOTM2N2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ezQ4M72yhVdHweKIe1JSppPxU25KTOqM11bZC6PHw2M"> Pre-Encoder Embeddings

@SGD2718 can you explain more about how you use fastconformer  to extract embedding ? Thank you so much !
